### DISCUSSION TITLE
Is there a way to do these things with nodes?

### DISCUSSION THREAD

**Discussion by Codesterz** [bnpr/Malt#161] [Q&A]
I only started messing with this render engine yesterday and have just been messing around with it to see what it can do. Now I'm not a programmer and what a lot of the nodes do I have no idea as the node names and functions seem to be geared more towards programmers and not artists. At the moment there are 3 things I'm trying to figure out how to do with the nodes. I see a whole bunch of complicated sounding nodes and thought maybe there is some sorta convoluted way to do the things I'm looking to do, but how to do it if it's possible is unknown to me.

1.  First off I'm assuming vec4 is a color node and assuming vec3 is a shader node. Is there a way to separate the colors from a vec4 node and then later rejoin them? Blender had nodes to do this Separate RGB, Combine RGB, Separate HSV, and Combine HSV.

2. Similar to the 1st question but instead of color is there a way to separate the X, Y, and Z information from something such as texture coordinates on a UV. Blender had the Separate XYZ, and Combine XYZ nodes for this.

3. Is there a way to adjust color values from a vec4 node such as the hue, saturation, brightness, contrast, and gamma. Blender had the nodes Hue/Saturation, Gamma, and Bright Contrast to achieve these.

These are the main things I was wondering how to do. If anyone knows of a way to do these with nodes the help would be much appreciated.

EDIT: I found vec4 split and vec4 join which lets me split and join the colors to RGBA, which just leaves the HSV parts of it unanswered.

**Comment by Codesterz** [comment]
![Node01](https://user-images.githubusercontent.com/52447070/141010725-3fee6487-f923-46fb-a171-bfef1fc552d1.jpg)
I'm gonna post this as a separate post instead of an edit since it'll be kinda long. So I THINK I found my answers to my 2nd question as I found a bunch of ivec, uvec, and bvec split/joins. I'm not to sure what each individual one is for at the moment, but I'll figure it out eventually. 

As I was figuring this stuff out though I realized a lot of the functions don't have a factor input. Upon realizing this I noticed I haven't found a node yet that lets me attach drivers to it. The model I've been working on was originally being made for eevee and used a toon shader I found and customized to look the way I wanted it. Eventually I ran into an issue with eevee due to the lack of light linking and started to look for an alternative which led me to trying out this render engine. 

My character model uses non spherical eyes so simply weighting a bone to the eye and changing the bones rotation to adjust where they are looking wasn't an option. Instead I had to move the texture around with bones using drivers and then to control the eye scaling I repositioned the texture placement to be centered with math nodes and had their scaling be based off the bone scaling. 

Is there a way to do this with  this render engine? The attached image shows the part of the nodes I was using in eevee to control the eyes position and scaling. Pay no mind to the clamps I accidentally clicked them when moving the nodes away from all the other nodes to make it more readable.

https://imgur.com/yFA4BWY

**Reply by pragma37** [reply]
I'm not sure why drivers are not available from node properties. I'll look into it.
For now, you can set up your drivers in the material panel.

Basic types in GLSL are:
bool: true or false values
float: Positive or negative number with decimals
int: Positive or negative number without decimals
uint: Positive number without decimals

Each basic type has its own vector type:
bvec: bool
vec: float
ivec: int
uvec: uint

https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Basic_types

**Comment by HSDSZ** [comment]
Vec3 and vec4 are both color. vec4 has an alpha channel.
rgb to hsv can be found in "common - color" functions

You are trying to use UV to move your eyes. Then use the node below (blender 2.93 + the latest malt development branch)
[example.zip](https://github.com/bnpr/Malt/files/7510449/example.zip)

**Reply by Codesterz** [reply]
I did see the RGB to HSV, but I couldn't figure out a way to separate the HSV since I'm mainly looking to adjust the 'V'(Value) without changing the Hue or Saturation. Subtracting then later adding to RGB results in a lot of color burning. I was trying to apply a texture to the pixel shader but I want the texture to be black(the subtracting) if it has no lighting. Then use the gradient diffuse to give it light(the adding) but it resulted in some harsh red in the mid lit area's. If I could do these to just the Value of HSV then I get the feeling the results would look closer to what I want. I'll have to look at your example later since I'm heading to bed, but nice to know it can be done.

**Reply by HSDSZ** [reply]
goodnight. check this example for splitting HSV
[New V.zip](https://github.com/bnpr/Malt/files/7510832/New.V.zip)

**Reply by pragma37** [reply]
vec3 just means a list of 3 values. It can be used to store positions, normals, RGB colors, HSV colors or anything else expressed as 3 numbers.

So for HSV editing, you can do rgb_to_hsv -> vec3_split -> do your edit -> vec3_join -> hsv_to_rgb.

You could also add a function to your custom lib to make it easier, for example:
```glsl
vec3 hsv_edit(vec3 rgb, float H, float S, float V)
{
    return hsv_to_rgb(rgb_to_hsv(rgb) + vec3(H,S,V));
}
````
